src/system.c

```diff
@@ -1519,8 +1512,50 @@ static void SySetGapRootPath( const Char * string )
         n++;
 #endif
     }
+    return;
 }
 
+/****************************************************************************
+**
+*F  SySetInitialGapRootPaths( <string> )  . . . . .  set the root directories
+**
+**  Setup GAP's initial root paths, based on two sources
+**  1) If the environment variable GAPROOT is undefined, the
+**     C macro GAPROOT_LOCATION (defined when GAP was compiled).
+**  2) Alternatively, try the location of GAP's executable.
+*/
+static void SySetInitialGapRootPaths(void)
+{
+    Int i, j;
+
+#ifdef GAPROOT_LOCATION
+    SySetGapRootPath(GAPROOT_LOCATION);
+    return;
+#endif
+
+    if (GAPExecLocation[0] != 0) {
+        /* GAPExecLocation might be a super-directory of GAP root,
+         * so we will go and search for the true GAP root.
+         * By default we try stepping back 2 levels */
+        char pathbuf[LONG_PATH_SIZE];
+        char initgbuf[LONG_PATH_SIZE];
+        for (i = 0; i < 3; ++i) {
+            strxcpy(pathbuf, GAPExecLocation, sizeof(pathbuf));
+            for (j = 0; j < i; ++j) {
+                strxcat(pathbuf, "../", sizeof(pathbuf));
+            }
```

Comment:
Wouldn't it be more natural to move `strxcpy(pathbuf, GAPExecLocation, ...`) outside the loop, and then do `strxcat(pathbuf, "../", ...` at the end of the loop?